lib/BackbeatConsumer.js

```diff
@@ -923,10 +918,21 @@ class BackbeatConsumer extends EventEmitter {
                 if (this._zookeeper) {
                     this._zookeeper.close();
                 }
-                if (this._consumer) {
+                if (this._consumer && this._consumer.isConnected()) {
                     this._consumer.unsubscribe();
-                    this._consumer.disconnect();
-                    this._consumer.on('disconnected', () => next());
+                    // Wait for partition unassign to complete before
+                    // disconnecting, the rebalance callback will handle
+                    // waiting for current jobs to complete as well as commit
+                    // the latest offsets
+                    this._waitForUnassign = true;
+                    const interval = setInterval(() => {
```

Comment:
why not "emit" a signal when unassigned, and simply waiting for that signal here?